File: ios/SiezaGym/Features/Home/SemanaCard.swift
import SwiftUI

/// La semana, igual a la de la web: siete días con flechas para moverse, el
/// día entrenado con el sólido del tema y hoy con anillo.
struct SemanaCard: View {
    @Environment(\.tema) private var tema
    let trainedDayKeys: Set<String>
    let streak: Int
    /// Cuántas semanas atrás se está mirando. 0 es la actual.
    @State private var offset = 0

    private var lunes: Date {
        let cal = TrainingCalendar.calendar
        let hoy = Date()
        let indice = TrainingCalendar.weekdayIndex(hoy)
        let base = cal.date(byAdding: .day, value: -indice, to: hoy) ?? hoy
        return cal.date(byAdding: .day, value: offset * 7, to: base) ?? base
    }

    private var dias: [(key: String, inicial: String, numero: Int, entrenado: Bool, esHoy: Bool, futuro: Bool)] {
        let cal = TrainingCalendar.calendar
        let hoyKey = TrainingCalendar.dayKey(Date())
        return (0..<7).compactMap { i in
            guard let fecha = cal.date(byAdding: .day, value: i, to: lunes) else { return nil }
            let key = TrainingCalendar.dayKey(fecha)
            return (key, HomeMetrics.dayLabels[i], cal.component(.day, from: fecha),
                    trainedDayKeys.contains(key), key == hoyKey, key > hoyKey)
        }
    }

    private var entrenadosEnSemana: Int { dias.filter(\.entrenado).count }

    var body: some View {
        GlassCard(padding: 16, radius: 26) {
            VStack(alignment: .leading, spacing: 14) {
                HStack(spacing: 8) {
                    flecha("chevron.left", habilitada: true) { offset -= 1 }
                    Text(rango)
                        .font(.system(size: 14, weight: .medium))
                        .foregroundStyle(tema.texto)
                        .frame(maxWidth: .infinity)
                    // No se puede haber entrenado en una semana que no pasó.
                    flecha("chevron.right", habilitada: offset < 0) { offset += 1 }
                }

                HStack(spacing: 4) {
                    ForEach(dias, id: \.key) { dia in
                        VStack(spacing: 7) {
                            Text(dia.inicial)
                                .font(.system(size: 10))
                                .foregroundStyle(tema.texto3)
                            Text("\(dia.numero)")
                                .font(.system(size: 13, weight: dia.entrenado ? .medium : .regular))
                                .foregroundStyle(color(dia))
                                .frame(width: 34, height: 34)
                                .background {
                                    if dia.entrenado { Circle().fill(tema.solido) }
                                }
                                .overlay {
                                    if dia.esHoy { Circle().strokeBorder(tema.texto, lineWidth: dia.entrenado ? 2 : 1) }
                                }
                                .opacity(dia.futuro ? 0.35 : 1)
                        }
                        .frame(maxWidth: .infinity)
                    }
                }

                Text(pie)
                    .font(.system(size: 11))
                    .foregroundStyle(tema.texto2)
            }
        }
        .animation(.smooth(duration: 0.2), value: offset)
    }

    private func color(_ dia: (key: String, inicial: String, numero: Int, entrenado: Bool, esHoy: Bool, futuro: Bool)) -> Color {
        // Hoy y entrenado a la vez: manda el color del relleno. Si ganara el de
        // "hoy" quedaría blanco sobre blanco.
        if dia.entrenado { return tema.sobreSolido }
        return dia.esHoy ? tema.texto : tema.texto2
    }

    private var rango: String {
        if offset == 0 { return "Esta semana" }
        let cal = TrainingCalendar.calendar
        let fin = cal.date(byAdding: .day, value: 6, to: lunes) ?? lunes
        let meses = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic"]
        let d1 = cal.component(.day, from: lunes), m1 = cal.component(.month, from: lunes)
        let d2 = cal.component(.day, from: fin), m2 = cal.component(.month, from: fin)
        return m1 == m2 ? "\(d1) – \(d2) \(meses[m1 - 1])" : "\(d1) \(meses[m1 - 1]) – \(d2) \(meses[m2 - 1])"
    }

    private var pie: String {
        guard entrenadosEnSemana > 0 else { return "Sin entrenamientos esta semana." }
        let base = "\(entrenadosEnSemana) \(entrenadosEnSemana == 1 ? "día entrenado" : "días entrenados")"
        guard offset == 0, streak > 0 else { return base }
        return "\(base) · 🔥 \(streak) \(streak == 1 ? "día seguido" : "días seguidos")"
    }

    private func flecha(_ icono: String, habilitada: Bool, accion: @escaping () -> Void) -> some View {
        Button(action: accion) {
            Image(systemName: icono)
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(tema.texto)
                .frame(width: 34, height: 34)
                .overlay { Circle().strokeBorder(tema.borde, lineWidth: 1) }
        }
        .disabled(!habilitada)
        .opacity(habilitada ? 1 : 0.3)
        .accessibilityLabel(icono == "chevron.left" ? "Semana anterior" : "Semana siguiente")
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/Features/Home/SemanaCard.kt
package com.siezagym.app.Features.Home

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowLeft
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.siezagym.app.DesignSystem.GlassCard
import com.siezagym.app.DesignSystem.LocalD2Theme
import com.siezagym.app.Domain.TrainingCalendar
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.ZonedDateTime

/** La semana, igual a la de la web: siete días con flechas para moverse, el
 *  día entrenado con el sólido del tema y hoy con anillo. */
@Composable
fun SemanaCard(
    entrenados: Set<String>,
    racha: Int,
    hoy: ZonedDateTime = ZonedDateTime.now(TrainingCalendar.zone),
) {
    val tema = LocalD2Theme.current
    var offset by remember { mutableIntStateOf(0) }

    val lunes = hoy.toLocalDate().minusDays((hoy.dayOfWeek.value - 1).toLong()).plusDays(offset * 7L)
    val hoyKey = TrainingCalendar.dayKey(hoy.toLocalDate())

    val dias = (0..6).map { i ->
        val fecha = lunes.plusDays(i.toLong())
        val key = TrainingCalendar.dayKey(fecha)
        DiaSemana(
            key = key,
            inicial = TrainingCalendar.dayLabels[i],
            numero = fecha.dayOfMonth,
            entrenado = entrenados.contains(key),
            esHoy = key == hoyKey,
            futuro = key > hoyKey,
        )
    }
    val entrenadosEnSemana = dias.count { it.entrenado }

    GlassCard(paddingInterno = 16.dp, radius = 26.dp) {
        Column {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Flecha(Icons.AutoMirrored.Filled.KeyboardArrowLeft, "Semana anterior", habilitada = true) { offset -= 1 }
                Text(
                    rango(lunes, offset),
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                    color = tema.texto,
                    modifier = Modifier.weight(1f),
                    textAlign = androidx.compose.ui.text.style.TextAlign.Center,
                )
                Flecha(Icons.AutoMirrored.Filled.KeyboardArrowRight, "Semana siguiente", habilitada = offset < 0) { offset += 1 }
            }

            Spacer(Modifier.height(14.dp))

            Row(horizontalArrangement = Arrangement.spacedBy(4.dp), modifier = Modifier.fillMaxWidth()) {
                dias.forEach { dia ->
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        modifier = Modifier.weight(1f),
                    ) {
                        Text(dia.inicial, fontSize = 10.sp, color = tema.texto3)
                        Spacer(Modifier.height(7.dp))
                        Box(
                            contentAlignment = Alignment.Center,
                            modifier = Modifier.size(34.dp),
                        ) {
                            if (dia.entrenado) {
                                Box(
                                    Modifier
                                        .size(34.dp)
                                        .clip(CircleShape)
                                        .background(tema.solido),
                                )
                            }
                            Text(
                                "${dia.numero}",
                                fontSize = 13.sp,
                                fontWeight = if (dia.entrenado) FontWeight.Medium else FontWeight.Normal,
                                color = color(dia, tema.solido, tema.sobreSolido, tema.texto, tema.texto2),
                            )
                            if (dia.esHoy) {
                                Box(
                                    Modifier
                                        .size(34.dp)
                                        .border(
                                            width = if (dia.entrenado) 2.dp else 1.dp,
                                            color = tema.texto,
                                            shape = CircleShape,
                                        ),
                                )
                            }
                        }
                    }
                }
            }

            Spacer(Modifier.height(14.dp))

            Row(verticalAlignment = Alignment.CenterVertically) {
                val base = if (entrenadosEnSemana == 0) {
                    "Sin entrenamientos esta semana."
                } else {
                    "${entrenadosEnSemana} ${if (entrenadosEnSemana == 1) "día entrenado" else "días entrenados"}"
                }
                if (entrenadosEnSemana > 0 && offset == 0 && racha > 0) {
                    Icon(
                        Icons.Filled.LocalFireDepartment,
                        contentDescription = null,
                        tint = tema.texto2,
                        modifier = Modifier.size(14.dp),
                    )
                    Spacer(Modifier.width(5.dp))
                    Text(
                        "$base · $racha ${if (racha == 1) "día seguido" else "días seguidos"}",
                        fontSize = 11.sp,
                        color = tema.texto2,
                    )
                } else {
                    Text(base, fontSize = 11.sp, color = tema.texto2)
                }
            }
        }
    }
}

private data class DiaSemana(
    val key: String,
    val inicial: String,
    val numero: Int,
    val entrenado: Boolean,
    val esHoy: Boolean,
    val futuro: Boolean,
)

private fun color(dia: DiaSemana, solido: Color, sobreSolido: Color, texto: Color, texto2: Color): Color {
    if (dia.entrenado) return sobreSolido
    if (dia.esHoy) return texto
    return if (dia.futuro) texto2.copy(alpha = 0.35f) else texto2
}

private fun rango(lunes: LocalDate, offset: Int): String {
    if (offset == 0) return "Esta semana"
    val fin = lunes.plusDays(6)
    val meses = listOf("ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic")
    return if (lunes.month == fin.month) {
        "${lunes.dayOfMonth} – ${fin.dayOfMonth} ${meses[lunes.monthValue - 1]}"
    } else {
        "${lunes.dayOfMonth} ${meses[lunes.monthValue - 1]} – ${fin.dayOfMonth} ${meses[fin.monthValue - 1]}"
    }
}

@Composable
private fun Flecha(
    icono: androidx.compose.ui.graphics.vector.ImageVector,
    descripcion: String,
    habilitada: Boolean,
    accion: () -> Unit,
) {
    val tema = LocalD2Theme.current
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(34.dp)
            .clip(CircleShape)
            .clickable(enabled = habilitada, onClick = accion)
            .border(1.dp, if (habilitada) tema.borde else Color.Transparent, CircleShape),
    ) {
        Icon(
            icono,
            contentDescription = descripcion,
            tint = if (habilitada) tema.texto else tema.texto.copy(alpha = 0.3f),
            modifier = Modifier.size(18.dp),
        )
    }
}